test/gpu/native/studies/transpose/transpose.chpl

```diff
@@ -3,15 +3,59 @@ use GpuDiagnostics;
 use GPU;
 use Time;
 
+enum implType {
+    naive, clever, lowlevel
+}
+use implType;
+
 config const perftest = false;
 config const numTrials = 10;
-config const useNaive = false;
+config const impl = naive;
 config const sizeX = 2048*8,
              sizeY = 2048*8;
 config param blockSize = 16;
 config param blockPadding = 1;
 config type dataType = real(32);
 
+inline proc transposeNaive(original, output) {
+  foreach (x,y) in original.domain do output[y,x] = original[x,y];
+}
+
+inline proc transposeClever(original, output) {
+  foreach 0..<original.size {
+    assertOnGpu();
+    setBlockSize(blockSize * blockSize);
+    param paddedBlockSize = blockSize + blockPadding;
+    var smArrPtr = createSharedArray(dataType, paddedBlockSize*blockSize);
+
+    const blockIdx = __primitive("gpu blockIdx x"),
```

Comment:
I feel like if we gave the loop an index variable we should be able to do a little bit of math to get the 2D index without calling any primitives.

My math might be off but assuming row-major order wouldn't it look something like this:

```
const idxX = idx % sizeX,
      idxY = idx / sizeX;
```

And then once we have the 2D index we can use that to calculate the block index and thread index:

```
blockIdxX = idxX / blockSize;
blockIdxY = idxY / blockSize;
threadIdxX = idxX % blockSize;
threadIdxY = idxY % blockSize;
```

Reply:
Computing indices like this doesn't quite work because it makes threads that aren't in the same block logically rely on each other's memory. I can compute the block index and thread index without primitives, but then I feel like I'd be doing the exact same logic that is "hidden" by the primitives themselves.

Reply:
> Computing indices like this doesn't quite work because it makes threads that aren't in the same block logically rely on each other's memory

I'm not sure I understand your comment (or maybe I'm misunderstanding what's being calculated?) I was suggesting you compute `blockIdxX`, `blockIdxY`, `threadIdxX`, `threadIdxY`, `idxX`, `idxY` in a way where you'd get the same values for a given loop iteration that you would with your existing approach; just without using our primitives.

> then I feel like I'd be doing the exact same logic that is "hidden" by the primitives themselves.

Right, I think the idea is to make the code use as few non user-facing features as possible (so we can say to our users: here's how you can do this type of computation in Chapel today using the documented features we have for native GPU support).  The fact that it's a bit of a hassle to do this gives us more motivation for finding a better way.

If you still want to let them know about the primitive we could say this in comments: e.g.: we don't have a user facing way to get thread ID, we have this primitive is available to do but this isn't meant to be a user-facing feature and we will come up with a user-facing way to do this more directly in the future.

You could also do it the reverse way (keep the primitive and explain how users could do it without the primitive if they wanted).

My preference would be to keep the code in `studies/` accurate to what a user could write today using documented features, the better; but I could be convinced otherwise if there's a compelling reason (it has a huge impact on performance or readability of the code or something).